8344911: Introduce test utility for asserting file open status

eirbjo opened the pull request:
Please review this PR which  adds a utility API in the test libraries to assert whether a file is currently open.

Several OpenJDK  tests currently rely on approximations to check this, including deletion (fails only on Windows), or checking `/proc/<pid>/fd` (Works only on Linux). These approximations are blunt instruments, it would be better to have an exact cross platform solution for this need to check if a file is open or closed.

Verification: With `OpenFilesTest` temporarily in tier1, GHA completed successfully across platforms.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344911](https://bugs.openjdk.org/browse/JDK-8344911): Introduce test utility for asserting file open status (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22343/head:pull/22343` \
`$ git checkout pull/22343`

Update a local copy of the PR: \
`$ git checkout pull/22343` \
`$ git pull https://git.openjdk.org/jdk.git pull/22343/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22343`

View PR using the GUI difftool: \
`$ git pr show -t 22343`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22343.diff">https://git.openjdk.org/jdk/pull/22343.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22343#issuecomment-2498489571)
</details>

RogerRiggs commented:
To the motivation, please list the tests that would be modified to use this test feature.
To the feature itself, it seems pretty complicated and invasive, please expand on the necessity.

Note that the initial comment in any PR is copied in every email sent on the PR and makes it harder to get to the point in subsequent emails. Additional details can/should be included in separate comments or in the Jira issue.

To have a clean PR, it would have been a benefit to reviewers to have squashed the work in progress commits before opening the PR.

Thanks for developing this approach to a difficult test problem.

eirbjo commented:
Moving details out of the PR description to reduce mail noise:

Initial support includes:

* `FileInputStream`, `FileOutputStream` and `RandomAccessFile` from `java.io`
* `FileChannelImpl` from `java.nio`, meaning we can check synchronous NIO APIs such as `Files.newInputStream`, `Files.newOutputStream`, `Files.newByteChannel` and `FileChannel.open`

This allows testing of many (most?) JDK APIs such as `ZipFile`, `JarFile`, `JarURLConnection`, `FileURLConnection` etc.

Tests use the following configuration to enable  instrumentation of file access APIs:

```
* @library /test/lib
* @run driver jdk.test.lib.util.OpenFilesAgent
* @run junit/othervm -javaagent:OpenFilesAgent.jar OpenFilesTest
```

Also provided is an assertion API to verify whether files are open or not:

```java
Path file = Files.createFile(Path.of("file.txt"));
assertClosed(file);
try (var is = Files.newInputStream(file)) {
    assertOpen(file);
}
assertClosed(file);
``` 

A failed `assertClosed` shows the stack trace for whatever action opened the file:

```
Caused by: java.lang.Throwable: Opening stack trace of jdk/build/macosx-x86_64-server-release/JTwork/scratch/file.txt
	at jdk.test.lib.util.OpenFiles.openFile(OpenFiles.java:59)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:357)
	at java.base/java.nio.file.Files.newByteChannel(Files.java:399)
	at java.base/java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:371)
	at java.base/java.nio.file.Files.newInputStream(Files.java:154)
	at OpenFilesTest.sample(OpenFilesTest.java:81)
```

See the library implementation test `OpenFilesTest` for samples.

eirbjo commented:
> To the motivation, please list the tests that would be modified to use this test feature. 

Thanks Roges! Yes, I realize this came a bit out of the blue.

The test I have been looking at so far include:

`test/jdk/java/net/URLClassLoader/RemoveJar.java`
This verifies that `URLClassLoader` (via JarURLConnection) when closed also closes the opened JarFile, under various loading scenarios and error conditions. The assertion that the file is closed is done by deleting the file and expecting that to fail. This expectation fails only on Windows.

`test/jdk/java/net/URLClassLoader/JarLoaderCloseTest.java`
Verifies that malformed URLs in the `Class-Path:" Manifest attribute of a JAR does not cause a file descriptor leak for the JAR file. Also relies on delete failing on Windows.

`test/jdk/sun/net/www/protocol/jar/FileURLConnectionLeak.java`
This verifies that calling `URLConnection::getLastModified` on a `JarURLConnection` does not leak a file descriptor for the underlying FileURLConnection. Also expects Files.delete to fail on Windows. On Linux, it checks files in `/proc/<pid>/fd`, which means it can only assert a closed file, not an open file.

I haven't looked too closely outside the java/net area, but expect there are other existing tests which could benefit from this. Perhaps there are also current or future tests which would have asserted file open status if a solution was available.

> To the feature itself, it seems pretty complicated and invasive, please expand on the necessity.

The goal is to:

* Make tests express their intent more clearly. The "delete on windows" idiom requires some explaining via comments. `assertClosed(file)` is more succinct and says what it does.
* Allow asserting open status without deleting the file. The need to delte means the assertion can only happen at the end of the test. With `assertOpen` / `assertClosed`, we can do the assert at any point in the test, meaning we can verify that an action opens or closes a file with a much finer granularity.
* Allowing asserts to work cross platform. This improves developer ergonomics, since a test can be developed without the need to test every change on multiple OS platforms (typically requiring a CI run).  

> Note that the initial comment in any PR is copied in every email sent on the PR and makes it harder to get to the point in subsequent emails. Additional details can/should be included in separate comments or in the Jira issue.

Good point, I moved the details into s separate comment.

> To have a clean PR, it would have been a benefit to reviewers to have squashed the work in progress commits before opening the PR.

Thanks, I squashed and force pushed before any code review comments. Hope that helps!

> Thanks for developing this approach to a difficult test problem.

Thanks for your comments. I know Java agents introduces some complexity, but I was positively surprised by the small number of classes needing instrumentation (just four). These are classes expected to have little churn.  I also found the new ClassFile API a pleasure to use.

eirbjo commented:
> To the feature itself, it seems pretty complicated and invasive

As for complexity, the feature consists of:

* A Java agent `OpenFilesAgent`, containing 277 source code lines with:
  * A `main` method to package itself as a JAR for use with `-javaagent` on the test's command line
  * A `premain` method to add the `ClassFileTransformer`, make the runtime class `OpenFiles` available for the JDK's boot class loader and retransform any classes if needed.
  * The class file transformer itself, injecting hooks at around open and close operations in `FileInputStream`, `FileOutputStream`,  `RandomAccessFile` and `sun.nio.ch.FileChannelImpl`. This uses the `java.lang.classfile` API.
* An API support class `OpenFiles`, containing 79 source code lines which:
  * Allows instumented code to register a file as being opened  or closed. 
  * Provides an API for tests to assert files as open or closed
* A test for the implementation the the feature itself, verifying that the feature works as expected across a wide range of JDK file primitives. 

To make use of the feature, tests must:

* Add a `@run driver` tag to prepare the agent JAR files
* Run in `/othervm`
* Add the `-javaagent:OpenFilesAgent.jar` JVM option
* Import the `OpenFiles` API
* Add `assertOpen`, `assertClosed` calls as required by the test logic
* Remove / clean up any old assertion mechanisms (if so desired) 
 
Yes, the feature has a non-negligble maintenence cost. However, given the potential benefits of cathing bugs early, I think the implementation size or compexity is not too bad considering what is offered.

Byte code instrumentation is never pretty, but with the new `ClassFile` API I think the resulting code was relatively straightforward. (Although I have used ASM a lot in the past, this was my first experience using the ClassFile API. I'm sure my code can be improved). And with the `ClassFile` API we can hopefully expect more maintainers to be experienced with class file transformations in the future.

eirbjo commented:
Here's my take on future maintenance costs of the utility:

Instrumentation churn is unlikely:

* JDK-8054720 introduced Java wrappers around native `open` methods in `FileInputStream`, `FileOutputStream` and `RandomFileAccess`, specifically to ease instrumentation. These are unlikely to change very often, if at all, same reasoning applies to the `close` methods.
* The signature of `FileChannelImpl:open` was last changed in 2014, when  JDK-8033917 added include the file path as a parameter. Before that, the signature did not change since the initial load.
* The `java.lang.classfile` API is exiting preview, so we can expect less churn in this area.

Some enhancement to the assertion API is likely:

* The initial assertion API is intentionally kept very limited, allowing only assertion of the open status of single files.
* A reasonable future enhancement would be to add support for running code in a block and asserting on the file actions occurring in the block as a whole. 
* One example could be `assertNoLeak` which takes a block of code and verifies that any files opened during the execution of the block is also closed in the block.
* Such APIs could also confine the scope of file access to the scope of the code block, preventing leaks in one test from effecting unrelated tests in the same JVM instance.

Feature creep is possible, but discouraged

* It is conceivable that maintainers would find it convenient to add instrumentation to this utility beyond file access, making t a more general tool for verifying various APIs. I'm not sure that's a good idea, such features should be evaluated and developed on their own merits and independently of this utility.     

Some bug fixing is likely

* The implementation test should verify most APIs / features, but yes, all code has bug. Assuming 1-25 bugs per 1000 lines of code, the implementation should have beween 0.3 and 9 bugs :-)

eirbjo commented:
Alright, after trying to use this utility for writing some more actual tests, I think it is a bit immature to consider this for integration in the current state.

While answering the question "is this particular file open at this exact moment?" is useful, being able to reason about files opened or closed within some code block/scope seems a lot more powerful and attractive. It would also make the utility less intrusive to unrelated code in the instrumented JVM and specifically it would isolate different tests opening the same files. 

Future work may be explored in this branch. Stay tuned :-)
